### ISSUE TITLE
HeadObject validation does not honor all validation parameters

### ISSUE THREAD

**Issue by naaaargle** [spulec/moto#2705]
Hi! This issue appears related to https://github.com/spulec/moto/issues/2263

**Observed Behavior**
I'm using `unittest` to write a test that will ensure validation on HeadObject calls work as expected. My `setUp` function includes the following code:
```python3.7
TEST_BUCKET = "fake_bucket"
TEST_KEY = "fake.csv"

@mock_s3
class TestVerifyS3Object(unittest.TestCase):
    def setUp(self):
        # Some other code ...
        self.s3_client = boto3.client("s3", region_name="us-east-1")
        self.s3_client.create_bucket(Bucket=TEST_BUCKET)
        self.s3_client.upload_file(Filename=tmp_file, Bucket=TEST_BUCKET, Key=TEST_KEY)
```

My unit tests have attempted both of the following:
```python3.7
        with self.assertRaises(botocore.exceptions.ClientError):
            self.s3_client.head_object(Bucket=TEST_BUCKET, Key=TEST_KEY, IfMatch="foo")
```
and
```python3.7
        with self.assertRaises(botocore.exceptions.ClientError):
            self.s3_client.head_object(
                Bucket=TEST_BUCKET, Key=TEST_KEY, IfUnmodifiedSince=datetime.datetime.utcnow())
```
Both result in `AssertionError: ClientError not raised`
I am invoking the tests locally using `nose==1.3.7`, and using `moto==1.3.14` which I installed using `pip`.

**Expected Behavior**
When I run the function I'm testing in a real environment using `boto==2.49.0`, `boto3==1.11.8`, and `botocore==1.14.9`, this code raises a `botocore.exceptions.ClientError`:
```python3.7
In [17]: try:
    ...:     client.head_object(
    ...:         Bucket=TEST_BUCKET, Key=TEST_KEY, IfMatch="foo",
    ...:     )
    ...: except botocore.exceptions.ClientError as e:
    ...:     print(e)
    ...:
An error occurred (412) when calling the HeadObject operation: Precondition Failed
```
This error response also happens when I execute this code in a lambda.

**Suspected Reason**
I did a little digging and have one idea as to why. I am admittedly not too familiar with this codebase but I hope this is helpful: 

The [`_key_response_head`](https://github.com/spulec/moto/blob/master/moto/s3/responses.py#L1114) method only looks at the `If-Modified-Since` header in the request, when actually there are [several validation headers](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.head_object) that may be passed to the `head_object` operation.

**If so...**
If this is the case, then I imagine something similar to https://github.com/spulec/moto/pull/2276/files would be necessary. If it is that simple, I'd be willing to take this on.

Thanks for reading!

**Comment by spulec** [comment]
Thanks for opening! I believe your diagnosis is correct and that we don't validate on any of the other headers.

If you get a chance, a PR would be great!

**Comment by bblommers** [comment]
This is now fixed as of moto >= 1.3.16.dev5
